src/drivers/ports/sam/i2c_port.i

```diff
@@ -0,0 +1,544 @@
+#define SAM_TWI_IER_ARBLST      (1 << 9)
+#define SAM_TWI_IER_SCLWS       (1 << 10)
+#define SAM_TWI_IER_EOSACC      (1 << 11) 
+#define SAM_TWI_IER_ENDRX       (1 << 12)
+#define SAM_TWI_IER_ENDTX       (1 << 13)
+#define SAM_TWI_IER_RXBUFF      (1 << 14)
+#define SAM_TWI_IER_TXBUFF      (1 << 15)
+
+
+/* TWI_IDR - Interrupt Disable Register */
+#define SAM_TWI_IDR_TXCOMP      (1)
+#define SAM_TWI_IDR_RXRDY       (1 << 1)
+#define SAM_TWI_IDR_TXRDY       (1 << 2)
+
+#define SAM_TWI_IDR_SVACC       (1 << 4)
+#define SAM_TWI_IDR_GACC        (1 << 5)
+#define SAM_TWI_IDR_OVRE        (1 << 6)
+
+#define SAM_TWI_IDR_NACK        (1 << 8)
+#define SAM_TWI_IDR_ARBLST      (1 << 9)
+#define SAM_TWI_IDR_SCLWS       (1 << 10)
+#define SAM_TWI_IDR_EOSACC      (1 << 11) 
+#define SAM_TWI_IDR_ENDRX       (1 << 12)
+#define SAM_TWI_IDR_ENDTX       (1 << 13)
+#define SAM_TWI_IDR_RXBUFF      (1 << 14)
+#define SAM_TWI_IDR_TXBUFF      (1 << 15)
+
+/* TWI_IMR - Interrupt Mask Register */
+#define SAM_TWI_IMR_TXCOMP      (1)
+#define SAM_TWI_IMR_RXRDY       (1 << 1)
+#define SAM_TWI_IMR_TXRDY       (1 << 2)
+
+#define SAM_TWI_IMR_SVACC       (1 << 4)
+#define SAM_TWI_IMR_GACC        (1 << 5)
+#define SAM_TWI_IMR_OVRE        (1 << 6)
+
+#define SAM_TWI_IMR_NACK        (1 << 8)
+#define SAM_TWI_IMR_ARBLST      (1 << 9)
+#define SAM_TWI_IMR_SCLWS       (1 << 10)
+#define SAM_TWI_IMR_EOSACC      (1 << 11) 
+#define SAM_TWI_IMR_ENDRX       (1 << 12)
+#define SAM_TWI_IMR_ENDTX       (1 << 13)
+#define SAM_TWI_IMR_RXBUFF      (1 << 14)
+#define SAM_TWI_IMR_TXBUFF      (1 << 15)
+
+/* TWI_RHR - RXDATA: Master or Slave Receive Holding Data*/
+#define SAM_TWI_RHR_MASK        (0xff)
+
+/* TWI_THR - TXDATA: Master or Slave Transmit Holding Data*/
+#define SAM_TWI_THR_MASK        (0xff)
+
+
+int i2c_port_module_init()
+{
+    return (0);
+}
+
+int i2c_port_init(struct i2c_driver_t *self_p,
+                  struct i2c_device_t *dev_p,
+                  int baudrate,
+                  int address)
+{
+    self_p->dev_p = dev_p;
+    self_p->twbr = baudrate;
+    self_p->address = address; /* Perhaps only used for i2c slave mode? */
+    return (0);
+}
+
+/* Example: i2c_port_set_clock( SAM_TWI1, 100000, 84000000); // 100KHz i2c clock, 84MHz CPU Clock */
+static void i2c_port_set_clock( struct i2c_device_t *dev_p, uint32_t dwTwCk, uint32_t dwMCk )
+{
+    uint32_t dwCkDiv = 0 ;
+    uint32_t dwClDiv ;
+    uint32_t dwOk = 0 ;
+
+    /* Configure clock */
+    while ( !dwOk )
+    {
+        dwClDiv = ((dwMCk / (2 * dwTwCk)) - 4) / (1<<dwCkDiv) ;
+
+        if ( dwClDiv <= 255 )
+        {
+            dwOk = 1 ;
+        }
+        else
+        {
+            dwCkDiv++ ;
+        }
+    }
+
+    /* Typical Arduino Due: 100KHz => CKDIV = 1 and CLDIV/CHDIV = 208 */
+    /* std_printf( FSTR("Using CKDIV = %u and CLDIV/CHDIV = %u\n\r"), dwCkDiv, dwClDiv ) ; */
+
+    dev_p->regs_p->CWGR = 0 ;
+    dev_p->regs_p->CWGR = (dwCkDiv << 16) | (dwClDiv << 8) | dwClDiv ;
+}
```

Comment:
This function looks like a copy of the Ateml, or possibly Arduino, implementation. Clarify under which licenses this code is. I will not accept code under any license but MIT.